src/main/kotlin/SquashingExporterHandler.kt

```diff
@@ -21,21 +21,35 @@
 
 package com.spotify.tracing
 
+import com.github.benmanes.caffeine.cache.Caffeine
+import com.github.benmanes.caffeine.cache.RemovalCause
 import io.opencensus.common.Timestamp
 import io.opencensus.trace.AttributeValue.booleanAttributeValue
 import io.opencensus.trace.AttributeValue.doubleAttributeValue
 import io.opencensus.trace.SpanId
+import io.opencensus.trace.TraceId
 import io.opencensus.trace.export.SpanData
 import io.opencensus.trace.export.SpanExporter
 import kotlinx.coroutines.GlobalScope
 import kotlinx.coroutines.launch
-
+import java.util.concurrent.TimeUnit
 
 class SquashingExporterHandler(
     private val delegate: SpanExporter.Handler,
     private val threshold: Int,
     private val whitelist: List<String>? = null
 ): SpanExporter.Handler() {
+    val cache = Caffeine.newBuilder()
+        .weakKeys()
+        .maximumSize(1_000)
+        .expireAfterAccess(2, TimeUnit.MINUTES)
+        .removalListener { traceId: TraceId?, spans: MutableList<SpanData>?, cause ->
+            if (cause == RemovalCause.EXPIRED || cause == RemovalCause.COLLECTED || cause == RemovalCause.SIZE) {
+                delegate.export(spans)
```

Comment:
i'm trying to understand what this does... does it flush any non squashed spans after 2 minutes or 1k spans are in the cache?

Reply:
Yes, or if the TraceId gets garbage collected. The `.weakKeys()` method on the builder causes it to use a weak reference, so if nothing else on the JVM has a strong reference it can get GCed. All 3 of those cases are safety nets in case a trace isn't working correctly.

Reply:
The reason for the if statement is that the listener gets called for every removal, including replacements or intentional invalidations.

Reply:
👍